File: examples/python/07-forms-annotations/main.py
# Extract form fields and annotations from a PDF.
# Run: python main.py form.pdf

import sys

from pdf_oxide import PdfDocument


def main():
    if len(sys.argv) < 2:
        print("Usage: python main.py <form.pdf>")
        sys.exit(1)

    path = sys.argv[1]
    doc = PdfDocument(path)
    print(f"Opened: {path}")

    for page in range(doc.page_count):
        # Form fields
        fields = doc.get_form_fields(page)
        if fields:
            print(f"\n--- Form Fields (page {page + 1}) ---")
            for f in fields:
                print(f'  Name: {f.name!r:<20} Type: {f.type:<12} '
                      f'Value: {f.value!r:<16} Required: {f.required}')

        # Annotations
        annotations = doc.get_annotations(page)
        if annotations:
            print(f"\n--- Annotations (page {page + 1}) ---")
            for a in annotations:
                print(f'  Type: {a.type:<14} Page: {page + 1}   '
                      f'Contents: "{a.contents or ""}"')

if __name__ == "__main__":
    main()

File: examples/rust/07-forms-annotations/main.rs
// Extract form fields and annotations from a PDF.
// Run: cargo run --example forms_annotations -- form.pdf

use pdf_oxide::PdfDocument;
use std::env;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let path = env::args().nth(1).expect("Usage: forms_annotations <file.pdf>");
    let doc = PdfDocument::open(&path)?;
    println!("Opened: {}", path);

    let page_count = doc.page_count();
    for page in 0..page_count {
        // Form fields
        let fields = doc.get_form_fields(page)?;
        if !fields.is_empty() {
            println!("\n--- Form Fields (page {}) ---", page + 1);
            for f in &fields {
                println!(
                    "  Name: {:<20} Type: {:<12} Value: {:<16} Required: {}",
                    format!("\"{}\"", f.name),
                    format!("{:?}", f.field_type),
                    format!("\"{}\"", f.value),
                    f.required
                );
            }
        }

        // Annotations
        let annotations = doc.get_annotations(page)?;
        if !annotations.is_empty() {
            println!("\n--- Annotations (page {}) ---", page + 1);
            for a in &annotations {
                println!(
                    "  Type: {:<14} Page: {}   Contents: \"{}\"",
                    format!("{:?}", a.annotation_type),
                    page + 1,
                    a.contents.as_deref().unwrap_or("")
                );
            }
        }
    }

    Ok(())
}
